### PR TITLE
SSL_OP_DISABLE_TLSEXT_CA_NAMES option implementation

### PR DESCRIPTION
**Pull request by NMorozxov**
##### Checklist

- [x] documentation is added or updated
- [x] tests are added or updated

### CODE REVIEW

File: doc/man3/SSL_CTX_set_options.pod [hunk 1]
```diff
@@ -67,6 +67,12 @@ The following B<bug workaround> options are available:
 Don't prefer ECDHE-ECDSA ciphers when the client appears to be Safari on OS X.
 OS X 10.8..10.8.3 has broken support for ECDHE-ECDSA ciphers.
 
+=item SSL_OP_DISABLE_TLSEXT_CA_NAMES
+
+Disable TLS Extention CA Names. You can want to disable it for security reasons.
+Another usage: some Windows TLS Intplementation crushes if this extention is
+bigger than 1024 bytes.
```
Comment (beldmit) [inline doc/man3/SSL_CTX_set_options.pod:74]: Better wording is smth like

Disable TLS Extension CA Names. You may want to disable it for security reasons or for compatibility with some Windows TLS implementations crashing when this extension is larger than 1024 bytes.

Comment (NMorozxov) [reply doc/man3/SSL_CTX_set_options.pod:74]: fixed

**Comment by NMorozxov** [comment]
@beldmit fix commited

**Comment by beldmit** [comment]
Travis failures seem irrelevant. 
Could you please take a look whether similar options are tested and, if yes, add a test for your option?

**Comment by NMorozxov** [comment]
Ok i will try
> Travis failures seem irrelevant.
> Could you please take a look whether similar options are tested and, if yes, add a test for your option?

**Comment by richsalz** [comment]
Does this need to be exposed at the s_client, etc., level?  Via SSL_conf_cmd or similar?

**Comment by NMorozxov** [comment]
SSL_CTX_set_options(ctx, SSL_OP_DISABLE_TLSEXT_CA_NAMES);
when creating SSL Context
> Does this need to be exposed at the s_client, etc., level? Via SSL_conf_cmd or similar?

**Comment by richsalz** [comment]
Can you use s_client to set this new option?  I do not think so.  I think it should be possible to use s_client to set this option. The way to do that is to add it to the tables (and documentation) for ```SSL_conf_cmd()/SSL_CTX_conf_cmd()```

**Comment by NMorozxov** [comment]
Ok i understand. Fixed
> Can you use s_client to set this new option? I do not think so. I think it should be possible to use s_client to set this option. The way to do that is to add it to the tables (and documentation) for `SSL_conf_cmd()/SSL_CTX_conf_cmd()`

**Comment by NMorozxov** [comment]
@beldmit @richsalz fix commited

**Comment by NMorozxov** [comment]
**EVP reason code 208 collision at EVP_R_SET_DEFAULT_PROPERTY_FAILURE**
is relevant but doen't belong to my PR
i just pull a clean repo and easy got this error after make update
Some ones other PR broke this

**Comment by NMorozxov** [comment]
on the fixing PR
https://github.com/openssl/openssl/pull/11708

**Comment by kaduk** [comment]
I would be interested in hearing more about the scenarios where this is seen to be needed.  While I do recall the stories of misconfigured systems that send out (e.g.) all of /etc/ssl/certs using lots of bandwidth, it's not immediately clear to me that an SSL option is better than just correcting the configuration to only send a few (or zero) certificates at all.
In general, I am reluctant to add configuration options to specifically disable a given TLS extension, since the TLS ecosystem is supposed to be robust to unknown extensions and we have other ways to influence the contents of this one in particular.  (It was, unfortunately, necessary to add options for extended-master-secret and encrypt-then-mac because of deployed implementations that choke badly when they receive them.  I'm not aware of such implementation bugs surrounding the certificate-authorities extension (hence my desire to know more).

**Comment by NMorozxov** [comment]
Imagine this scenario: I want to make a web server with TLS authentication using GOST12-GOST89-GOST8912 cjipersuite, that allowed ONLY for users from Russian government certificated CA. Check the users after TLS connection established is prohibited by  moderator.
The solution:
1) I can build nginx http://nginx.org/ with OpenSSL
2) I can download the CA certs from http://e-trust.gosuslugi.ru/CA/DownloadTSL?schemaVersion=0
3) I add all certificates in nginx ssl_trusted_certificate and set ssl_verify_client on
4) I works, but some Windows can't connect. Disabling this extention will allow them to connect.

**Comment by NMorozxov** [comment]
ping @beldmit @kaduk

**Comment by beldmit** [comment]
I think it also worth adding an option to the `s_server` application.

**Comment by kaduk** [comment]
Thanks for the clear description of the problematic situation.  It seems like there would be a few other avenues that could (at a technical level) be tweaked to allow things to work, so I personally do not want to endorse this mechanism.  Having said that, I will not object if someone else approves it and it goes into the tree.

**Comment by beldmit** [comment]
@kaduk, I am not sure I understand which other avenues you mean.

**Comment by NMorozxov** [comment]
@beldmit option to s_server commited

File: apps/s_server.c [hunk 2]
```diff
@@ -952,6 +952,8 @@ const OPTIONS s_server_options[] = {
     {"anti_replay", OPT_ANTI_REPLAY, '-', "Switch on anti-replay protection (default)"},
     {"no_anti_replay", OPT_NO_ANTI_REPLAY, '-', "Switch off anti-replay protection"},
     {"http_server_binmode", OPT_HTTP_SERVER_BINMODE, '-', "opening files in binary mode when acting as http server (-WWW and -HTTP)"},
+    {"no-CANames", OPT_NOCANAMES, '-',
```
Comment (beldmit) [inline apps/s_server.c:955]: Not minus but underscore?

**Comment by NMorozxov** [comment]
@beldmit fix commited
but there is 
{"no-CAfile", OPT_NOCAFILE, '-',

**Comment by beldmit** [comment]
Travis failures are relevant.

**Comment by NMorozxov** [comment]
@beldmit fix commited

**Comment by kaduk** [comment]
> @kaduk, I am not sure I understand which other avenues you mean.

Mostly it relates to "Check the users after TLS connection established is prohibited by moderator." being a policy (non-technical) requirement.  I don't know why it's there, or whether it would also prohibit checking the users during (not after) TLS connection establishment.  Based on my current knowledge it appears as just an arbitrary unneeded restriction.
There's also an option to tell people with old, unsupported, non-protocol-compliant windows clients that they will need to find something that implements the spec properly, though I do not actually expect anyone to pick this route.
I also don't know the details of the certification process, who is eligible to get certified, how many stale entries there are in the presented corpus, or whether it's practical to enforce a consolidation to a small subset of the certified CAs.

File: apps/s_server.c (hunk 2 above)
Comment (t8m) [reply apps/s_server.c]: Also I'd put all in lower case. I do not think the CAfile no-CAfile example is a good thing to follow.

File: apps/s_server.c [hunk 3]
```diff
@@ -952,6 +952,8 @@ const OPTIONS s_server_options[] = {
     {"anti_replay", OPT_ANTI_REPLAY, '-', "Switch on anti-replay protection (default)"},
     {"no_anti_replay", OPT_NO_ANTI_REPLAY, '-', "Switch off anti-replay protection"},
     {"http_server_binmode", OPT_HTTP_SERVER_BINMODE, '-', "opening files in binary mode when acting as http server (-WWW and -HTTP)"},
+    {"no_CANames", OPT_NOCANAMES, '-',
```
Comment (t8m) [inline apps/s_server.c:955]: Use `no_ca_names` please. And please  rename the variable as well.

Comment (NMorozxov) [reply apps/s_server.c]: fixed

**Comment by NMorozxov** [comment]
@t8m fix commited

**Comment by t8m** [comment]
My approval is pending until the CI passes.

**Comment by NMorozxov** [comment]
Very strange travis job fails, can you restart it

**Comment by beldmit** [comment]
Travis is happy.

File: apps/s_server.c (hunk 2 above)
Comment (richsalz) [reply apps/s_server.c]: Using no-CA{file,path,store} is common -- 27 cases, so please use that convention.

Comment (beldmit) [reply apps/s_server.c]: Well, this option has a different scope so I'm not sure that a convention is relevant here.

Comment (richsalz) [reply apps/s_server.c]: you are correct and I was wrong.

**Comment by kaduk** [comment]
> > @kaduk, I am not sure I understand which other avenues you mean.
> 
> Mostly it relates to "Check the users after TLS connection established is prohibited by moderator." being a policy (non-technical) requirement. I don't know why it's there, or whether it would also prohibit checking the users during (not after) TLS connection establishment. Based on my current knowledge it appears as just an arbitrary unneeded restriction.
> There's also an option to tell people with old, unsupported, non-protocol-compliant windows clients that they will need to find something that implements the spec properly, though I do not actually expect anyone to pick this route.
> I also don't know the details of the certification process, who is eligible to get certified, how many stale entries there are in the presented corpus, or whether it's practical to enforce a consolidation to a small subset of the certified CAs.

In particular, it seems like it should be possible to have the initial SSL_CTX/SSL configuration not include CA names and only load the CA configuration during the verify callback.  This would result in no configured CAs while the CertificateRequest is being constructed (and so the extension would not be sent), but (IIUC) would still allow using the configured CAs in the vallidation process.  Yes, it's pretty messy, and would require a bit of care to avoid a lot of IO on each connection, but it seems technically possible.

**Comment by beldmit** [comment]
It seems to be an overengineering to me :(

**Comment by NMorozxov** [comment]
For me my option looks like a automation for this
> > > @kaduk, I am not sure I understand which other avenues you mean.
> > 
> > 
> > Mostly it relates to "Check the users after TLS connection established is prohibited by moderator." being a policy (non-technical) requirement. I don't know why it's there, or whether it would also prohibit checking the users during (not after) TLS connection establishment. Based on my current knowledge it appears as just an arbitrary unneeded restriction.
> > There's also an option to tell people with old, unsupported, non-protocol-compliant windows clients that they will need to find something that implements the spec properly, though I do not actually expect anyone to pick this route.
> > I also don't know the details of the certification process, who is eligible to get certified, how many stale entries there are in the presented corpus, or whether it's practical to enforce a consolidation to a small subset of the certified CAs.
> 
> In particular, it seems like it should be possible to have the initial SSL_CTX/SSL configuration not include CA names and only load the CA configuration during the verify callback. This would result in no configured CAs while the CertificateRequest is being constructed (and so the extension would not be sent), but (IIUC) would still allow using the configured CAs in the vallidation process. Yes, it's pretty messy, and would require a bit of care to avoid a lot of IO on each connection, but it seems technically possible.

**Comment by beldmit** [comment]
Merged. Thanks!

**Pull request closed without merging**